### PR TITLE
Feature/#50 Google Analytics

### PR DESCRIPTION
**Pull request by edward-hong**
# Details

## Description

The aim of this PR is to add Google Analytics. The following is a list of steps I took to achieve this.

1. Then I added some code to `scripts/custom-config.js`. This is so Webpack can compile the `process.env` variables so we can use them on the front end.
2. Created a file called `googleAnalytics.js` in the `src` directory that exports two functions. One initializes the Google Analytics and the second track the current page.
3. Rewrote `App.js` into a class component. The reason I did this was that I thought the best way to run the two functions was during `componentDidMount` life cycle hook.

~~Added `.env.example` file with a list of variables~~
~~Made a small amendment to README.md about `.env.example`~~

refer to #81 regarding `process.env`

## Issue

[#50](https://github.com/Coding-Coach/coding-coach/issues/50) Google Analytics

### CODE REVIEW

File: scripts/custom-config.js [hunk 1]
```diff
@@ -71,5 +73,18 @@ function rewireModule(modulePath, customizer) {
   // react-scripts imports the config as a `const` and we can't
   // modify that reference.
   let config = defaults.__get__('config');
+
+  // Have webpack compile the env keys so the can be used on the client
+  var env = dotenv.config().parsed;
```
Comment (linux-nerd) [inline scripts/custom-config.js]: @edward-hong CRA has inbuilt configuration for .env file. If you have your keys renamed to REACT_APP_GA_TRACKING_ID in .env file, you can access the same via process.env.REACT_APP_GA_TRACKING_ID from js.

File: src/googleAnalytics.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import ReactGA from 'react-ga';
+
+export const initializeGA = () => ReactGA.initialize(process.env.GA_TRACKING_ID);
+export const trackCurrentPage = () => ReactGA.pageview(window.location.pathname);
```
Comment (nitish24p) [inline src/googleAnalytics.js]: Ideally this track page should be called only if GA has been initialised

File: src/containers/App/App.js [hunk 3]
```diff
@@ -3,30 +3,37 @@ import { func } from 'prop-types';
...
-        >
-          {t('landing-page-start')}
-        </a>
-      </header>
-    </div>
-  );
+class App extends React.Component {
+  componentDidMount() {
+    initializeGA();
+    trackCurrentPage();
```
Comment (nitish24p) [inline src/containers/App/App.js]: Both These calls can be part of a separate file like analytics.js..
suppose tomorrow we want to include mixpanel, or heap or some other lib, we add that in the abstracted analytics.js and it gets applied everywhere where ever analytics.js has been invoked

File: src/googleAnalytics.js (hunk 2 above)
Comment (edward-hong) [reply src/googleAnalytics.js]: Yes I agree. I've read through react-ga docs again. There's no mention of a method that checks to see if GA is initialized. I guess instead of export const each function like I've done, I could export an object with these two functions as methods and a property called something like 'hasGABeenInit' which is a boolean and everytime the init function is called, 'hasGABeenInit' will then be set to true. Would that be the best approach?

File: src/containers/App/App.js (hunk 3 above)
Comment (edward-hong) [reply src/containers/App/App.js]: You've suggested to abstract the Google Analytics code. I thought that was a good idea so I abstracted it into a file called `googleAnalytics.js` in the `src` directory. I named the file name specifically enough to make sure that the code is for that one purpose. I'm not sure how the possibility of people wanting to add other libraries into the file that I made called `googleAnalytics.js` applies here.

File: scripts/custom-config.js (hunk 1 above)
Comment (edward-hong) [reply scripts/custom-config.js]: Thanks for the heads up! I'll make sure to make the adjustments you suggested.

File: src/googleAnalytics.js (hunk 2 above)
Comment (edward-hong) [reply src/googleAnalytics.js]: @nitish24p I've changed the code in `googleAnalytics.js` is this better?

```
export default {
  hasGABeenInit: false,
  initGA: () => {
    ReactGA.initialize(process.env.REACT_APP_GA_TRACKING_ID);
    this.hasGABeenInit = true;
  },
  trackCurrentPage: () => {
    if (this.hasGABeenInit) {
      ReactGA.pageview(window.location.pathname);
    }
  }
}
```

Comment (nitish24p) [reply src/googleAnalytics.js]: The `this` context here will be the global window object not the object.. The idea is good but remove all the arrow functions for object method declarations. You can read more here https://dmitripavlutin.com/when-not-to-use-arrow-functions-in-javascript/
```javascript
export default {
  hasGABeenInit: false,
  initGA: function() {
    ReactGA.initialize(process.env.REACT_APP_GA_TRACKING_ID);
    this.hasGABeenInit = true;
  },
  trackCurrentPage: function() {
    if (this.hasGABeenInit) {
      ReactGA.pageview(window.location.pathname);
    }
  }
}
```

File: src/googleAnalytics.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import ReactGA from 'react-ga';
+
+export default {
+  hasGABeenInit: false,
+  initGA: function() {
```
Comment (arthurbuhl) [inline src/googleAnalytics.js]: I think we can use a shorter syntax for method definitions on objects initializers. Given the following code:

```js
var obj = {
  foo: function() {
    /* code */
  },
  bar: function() {
    /* code */
  }
};
```

You are now able to shorten this to:
```js
var obj = {
  foo() {
    /* code */
  },
  bar() {
    /* code */
  }
};
```

MDN Doc Link: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions

File: README.md [hunk 5]
```diff
@@ -72,3 +72,9 @@ You might have some conflicts while rebasing, try to resolve the conflicts for e
 In order to merge a PR, it will first go through a review process. Once it is approved, we will merge to the `development` branch using the `Squash` button in github.
 
 When using squash, all the commits will be squashed into one. The idea is to merge features/fixes as oppose of merging each individual commit. This helps when looking back in time for changes in the code base, and if the PR has a great comment, it's easier to know why that code was introduced.
+
+### Google Analytics
+
+Create a `.env` file in the root directory. Add your Tracking ID under the variable `REACT_APP_GA_TRACKING_ID` e.g:
+
+    REACT_APP_GA_TRACKING_ID='UA-xxxxxxxxx-x'
```
Comment (arthurbuhl) [inline README.md]: I think we should add a sample `.env` like `.env.sample` with all the ids and commit to git but with empty or demo values. This is done so that all the devs are in sync about all the env used in the app. Or maybe we can create a separate markdown file mentioning about all the env variables (just the varibale names, not the key values) used in the app.

File: src/containers/App/App.js [hunk 6]
```diff
@@ -3,34 +3,37 @@ import { func } from 'prop-types';
...
-  );
+class App extends React.Component {
+  componentDidMount() {
+    GA.initGA();
+    GA.trackCurrentPage();
+  }
+  render() {
+    return (
+      <div className={styles.app}>
+        <img src={require('./assets/mentor.jpg')} alt="Mentor" className={styles.image} />
```
Comment (arthurbuhl) [inline src/containers/App/App.js]: I think we should use `i18n` for this static text used for image alt.

File: README.md (hunk 5 above)
Comment (edward-hong) [reply README.md]: Yea I think it's good for contributors to have some default keys to quickly get up and running

File: src/containers/App/App.js (hunk 6 above)
Comment (edward-hong) [reply src/containers/App/App.js]: This isn't relevant to this PR regarding Google Analytics and setup of `.env` as a consequence of GA

Comment (arthurbuhl) [reply src/containers/App/App.js]: Yes. I think I will create a new issue for it.

File: README.md (hunk 5 above)
Comment (arecvlohe) [reply README.md]: I think having an `.env.example` is a good idea so all the devs know what env vars there are

Comment (edward-hong) [reply README.md]: Ok I'll add an `.env.example` file with a GA Tracking so everyone can get up and running smoothly.

File: src/googleAnalytics.js (hunk 4 above)
Comment (edward-hong) [reply src/googleAnalytics.js]: @arecvlohe You're right. I'll update the PR for this.

File: src/googleAnalytics.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import ReactGA from 'react-ga';
```
[the remark this answers is not in the document]
Comment (edward-hong) [reply src/googleAnalytics.js]: I've moved the files to `src/utils/ga.js`

**Comment by arthurbuhl** [comment]
Any updates on this PR? Should we close it or keep it open for few more weeks?

**Comment by crysfel** [comment]
A lot of things have changed since this PR was open. Given that there's no activity here, I'm going to close it, as it will be easier to just install `react-ga` and do the setup than fixing all the conflicts xD
